README.md
FAST-iCLIP
==========

Fully Automated and Standardized iCLIP (FAST-iCLIP) is a fully automated tool to process iCLIP data. Please cite the following paper:

`Flynn RA, Martin L, Spitale RC, Do BT, Sagan SM, Zarnegar B, Qu K, Khavari PA, Quake SR, Sarnow P, Chang HY (2014). Dissecting non-coding and pathogen RNA-protein interactomes. RNA 21:1, 1-9. doi:10.1261/rna.047803.114`

This package contains two main sets of tools: an executable called `fasticlip` to run iCLIP on human and mouse data, and several iPython notebooks to process iCLIP data from viral genomes.

The following README will focus mainly on `fasticlip`. The pdf in the repository contains further instructions for using the iPython notebooks.

**Table of Contents**

- [Usage](#user-content-usage)
	- [Required arguments](#user-content-required-arguments)
	- [Optional arguments](#user-content-optional-arguments)
- [Installation instructions](#user-content-installation-instructions)
- [Dependencies](#user-content-dependencies)
- [Input](#user-content-input)
- [Output](#user-content-output)
- [How the pipeline works](#user-content-how-the-pipeline-works)

Usage
-----

`fasticlip [-h] -i INPUT [INPUT ...] [--trimmed] [--hg19 | --mm9] -s STAR_INDEX -n NAME -o OUTPUT [-f N] [-a ADAPTER] [-tr REPEAT_THRESHOLD_RULE] [-tn NONREPEAT_THRESHOLD_RULE] [-sr STAR_RATIO] [-bm BOWTIE_MAPQ] [-q Q] [-p P] [-l L] [--verbose]`

Example: `fasticlip -i rawdata/example_MMhur_R1.fastq rawdata/example_MMhur_R2.fastq --mm9 -s /seq/STAR/indexes/mm9 -n MMhur -o results`

### Required arguments

  flag | description
  ------------------|------------------------------------------------
  -h, --help   |      show this help message and exit
  -i INPUT(s) | At least one input FASTQ (or fastq.gz) files; separated by spaces
  --hg19        |    required if your CLIP is from human
  --mm9          |   required if your CLIP is from mouse
  -s STAR_INDEX  |   Path to STAR index for your organism
  -n NAME         |  Name of output directory
  -o OUTPUT        | Name of directory where output directory will be made
  
### Optional arguments

  flag | description
  ------------------|------------------------------------------------
  --trimmed       |      flag if files are already trimmed
  -f N          |    Number of bases to trim from 5' end of each read. Default is 14.
  -a ADAPTER     |   3' adapter to trim from the end of each read. Default is A            GATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATGCCGTCTTCTGCTTG.
  -tr REPEAT_THRESHOLD_RULE | m,n: at least m samples must each have at least n RT stops mapped to repeat RNAs. Default is 1,4 (1 sample); 2,3 (2 samples); x,2 (x>2 samples)
  -tn NONREPEAT_THRESHOLD_RULE | m,n: at least m samples must each have at least n RT stops mapped to nonrepeat RNAs. Default is 1,4 (1 sample); 2,3 (2 samples); x,2 (x>2 samples)
  -sr STAR_RATIO   |     Maximum mismatches per base allowed for STAR genome mapping (corresponds to outFilterMismatchNoverLmax). Default is 0.08 (2 mismatches per 25 mapped bases).
  -bm BOWTIE_MAPQ   |    Minimum MAPQ (Bowtie alignment to repeat/tRNA/retroviral indexes) score allowed. Default is 42.  
  -q Q         |         Minimum average quality score allowed during read filtering. Default is 25.
  -p P          |    Percentage of bases that must have quality > q during filtering. Default is 80.
  -l L           |       Minimum length of read. Default is 15.
  --clipper	| run CLIPper on data. Not recommended as we find that this removes a lot of true CLIP reads
  --verbose	| Prints out lots of things :)


Installation instructions
------------

1. Clone this repository by running one of the following:
	- `git clone git@github.com:ChangLab/FAST-iCLIP.git` if you use ssh authentication
	- `git clone https://github.com/ChangLab/FAST-iCLIP.git` otherwise
2. Type `cd FAST-iCLIP` to enter the folder.
3. Run `./configure`. This will check for dependencies (below) and download necessary files (bowtie indices, gene lists and genomes, and example iCLIP data).
4. Run `sudo python setup.py install`. If you do not have sudo privileges, run `python setup.py install --user` or `python setup.py install --prefix=<desired directory>`.
5. You should see three new folders inside `FAST-iCLIP`: `docs`, `rawdata`, and `results`.
6. Make your genome index for STAR. Once STAR is installed, run the following: `STAR --runMode genomeGenerate --runThreadN 8 --genomeDir <your desired folder> --genomeFastaFiles <genome FASTA file>`. You can download FASTA files from http://hgdownload.cse.ucsc.edu/downloads.html.
5. Try running the following command: 
  `fasticlip -i rawdata/example_MMhur_R1.fastq rawdata/example_MMhur_R2.fastq --mm9 -s <location of your STAR index> -n MMhur -o results`. It should run in ~1 hour. Look inside `results/MMhur` for output files.

You can `fasticlip` from outside its installation directory. To do this, add the following lines to the end of your `.bash_profile` script:

	export FASTICLIP_PATH=<your absolute FAST-iCLIP installation path>
	export PATH=$FASTICLIP_PATH:$PATH
Save the file, then run `source ~/.bash_profile`.

Dependencies
------------

1. Python 2.7: https://www.python.org/download/releases/2.7/
2. iPython: http://ipython.org/install.html (optional)
3. iPython notebook: http://ipython.org/notebook (optional)
4. Matplotlib for Python (plotting): http://matplotlib.org/
5. Pandas for Python (data): http://pandas.pydata.org/
6. Bowtie2: http://bowtie-bio.sourceforge.net/bowtie2/index.shtml
7. STAR: https://github.com/alexdobin/STAR
8. bedtools: http://bedtools.readthedocs.org/en/latest/
9. CLIPper: https://github.com/YeoLab/clipper/wiki/CLIPper-Home (now optional)
10. FASTX-Tookit: http://hannonlab.cshl.edu/fastx_toolkit/
11. Matplotlib-venn for Python: https://pypi.python.org/pypi/matplotlib-venn
12. iCLIPro: http://www.biolab.si/iCLIPro/doc/

Input
-----

At least one FASTQ or compressed FASTQ (fastq.gz). Use the `--trimmed` flag if trimming has already been done.

Output
------

Three subdirectories inside the named directory within `results`. 
- `figures` has 6 figures in pdf and png format.
  - *Figure 1* visualizes the some of the relevant summary data.
    - It includes a read count summary per pipeline step. The source data is: PlotData_ReadsPerPipeFile
    - It includes a pie chart of UTR binding.
    - This uses reads obtrained from intersection with ENSEMBL-derived UTR coordinates. 
    - The source data is: PlotData_ReadsPerGene_*UTR or CDS
    - It also includes a bar graph of gene count per RNA type.
    - The source data is: PlotData_ReadAndGeneCountsPerGenetype
 
  - *Figure 2* provides a richer summary of the UTR data.
    - The upper panel is an aggregate trace of binding derived from a custom perl script.
    - The lower panels provide a heatmap of binding intensity for gene exclusivly bound in each UTR or CDS.
    - This allows us to isolate genes with exclusive UTR,CDS,or intronic binding.
    - The source data is: PlotData_ExclusiveBound_*

  - *Figures 3 and 4* provide coverage histograms of binding across each repeat RNA and rRNA, respectivly. 
    - Source data: PlotData_RepeatRNAHist_*

  - *Figure 5* provides a summary of snoRNA binding data.
    - The pie chart provides a summary of reads per snoRNA type.
    - These are complimented by histograms of RT stop position within the snoRNA gene body.

  - *Figure 6* provides histograms of RT stop position within gene body for all remaining ncRNA types.

- `rawdata` has all the PlotData files used to make the figures, as well as intermediate files that can be useful in generating other plots.
- `todelete` has files that are unnecessary to keep.

How the pipeline works
----------------------

1. Prepare 2 FASTQ or FASTQ.gz files corresponding to the two replicates for the iCLIP experiment.

2. Duplicate removal, quality filter, and trim adapter from the 3' end from the reads
  - Duplicate removal:
	  - This step takes advantage of the fact that 5' end of each read has a random barcode.
	  - Each initial starting molecule that was RT'd will have a unique barcode.
	  - Therefore, PCR duplicates are removed by collapsing molecules with identical 5' barcode sequences.
  - RT primer is cleaved, leaving adapters. 
  - Remove adapter region from the 3' end of the read.
  - The adapter is an optional input parameter.
  - Default is to remove sequences less than N=33 nucleotides. 
  - Q33 specifies the quality score encoding format.

3. After duplicate removal, remove the 5' barcode sequence. Default is 13 nt removed.

4. We then map to a repeat index, then to a tRNA index, then to a genome index.
  - We only keep reads that are unique and perfectly aligned.
  - We then remove reads that map to blacklist or repeat regions.

5. After mapping, we isolate the 5' position (RT) stop for both positive and negative strand reads.
  - This represents the cross-link site in the initial experiment.

6. For each strand, we merge RT stops between replicates. 
  - This means that at RT stop position must be conserved between replicates.
  - If conserved, we count the total number of instances of the RT position for both replicates.
  - If the total counts exceed a specified threshold, then we record these RT stops.
  - Finally, we re-generate a "read" around the RT stop using the passed parameter "expand," 
  - A "read" around the RT stop is required for downstream processing.

7. Expanded reads from RT stop merging are passed to CLIPper, a peak calling algorithm.
  - CLIPper returns a bed-like file format with window coordinates, reads counted per window, etc.
  - We use these windows to extract "low FDR" reads from the total set of reads passed to CLIPper.
  - We then make bedGraph and BigWig files from this complete pool of "low FDR" reads, allowing easy visualization. 

8. Partition "low FDR" reads by gene type.
  - We partition the gene names recoved from CLIPper using ENSEMBL annotation of gene name by RNA type.
  - Once this is done, we also split the "low FDR" reads recovered from CLIPper by type using the gene name.
  - Protein coding and lincRNA genes can be embedded snoRNAs or miRNAs that make this more challenging.
  - In turn, we re-generate the initial RT stop and intersect this with two different filters.
  - One filter is a snoRNA mask and the other filter is a miR mask. 
  - Both are derived from Ensembl annotations.
  - These masks allow us to remove all "protein coding" RT stops that fall within annotated sno/mi-RNA regions.

9. Quantification of reads per gene.
  - For each gene type, we quantify the number of reads per gene.
  - For all but snoRNAs, this is computed using the bed files obtained above.
  - For snoRNAs, we intersect the initial pool of "low FDR" reads with custom annotation file. 
  - The custom annotation file is a summation of UCSC, Rfam, and ENSEMBL loci annotated as snoRNAs as no individual annotation entirely captured all known loci.
  - Collectively, this gives us reads per gene for each gene type.
  - All are based upon ENSEMBL annotation except for the snoRNAs.

10. Summary of RT stop intensity around CLIPper cluster centers.
  - We generate a bed file of cluster center positions using the CLIPper cluster file output.
  - We use a custom perl script that generates a heatmap of RT stop intensty per cluster.
  - This allows us to later visualize the distribution of RT stops per cluster.

11. Partition protein coding reads by UTR.
  - We intersect sno/mi-RNA filtered reads with ENSEMBL-derived UTR coordinates. 
  - We perform this such that each read assignment is mutually exclusive.
  - This only isolates reads that fall within each UTR type.
  - Similarly, we use a custom perl script to generate a matrix of read intensity per gene.
  - This provides a complete binding profile per gene.

12. Partition reads by ncRNA binding region.
  - For non-coding RNAs, we simply annotate reads with the start and stop position for each ncRNA.
  - This allows us to determine the position of each RT stop with respect to the full length of the gene.

13. Partition repeat-mapped RT stops by region.
  - The repeat RNA mapped RT stops are paritioned using the repeat custom index annotation. 
  - As with the ncRNAs, this is later used for visualization.


CHANGELOG.md
# Changelog

## Unreleased / Known issues
- can run the other scripts (viralgenome, motifAnalysis) outside the installation directory
- Need to move all global variables to cfg.py
- Add hg38 and mm10 annotation files
- docs.tar.gz needs to be updated
- Add more analysis of retroviral data (currently in another script)

## [0.9.3] - 2016-02-07
### Added
- New flag to control STAR N over L ratio (mismatches per base). Default set at 0.08

### Fixed
- Thresholding now implemented correctly for tRNA, repeat, and genome

## [0.9.2] - 2016-02-02
### Added
- Maps to retroviral index now

### Fixed
- ReadsPerPipeFile now outputs counts

## [0.9.1] - 2016-01-31
### Fixed
- Script runs to completion with --clipper
- Figure 1 is no longer stretched out vertically
- When not using CLIPper, protein-coding genes are now correctly enumerated
- Folder structure is now correct

## [0.9.0] - 2016-01-29
### Added
- Installation procedure using setup.py
- Now runs STAR instead of bowtie2 for genome index mapping. STAR is faster and maps across splice junctions.
- Users now have to supply a STAR index (-s flag required)
- CLIPper is now optional (and no longer recommended). Instead, reads are now overlapped by snoRNAs and miRNAs before overlapped to genes.
- Logging is improved, with a --verbose flag to see all command lines run
- Histograms of tRNA isotypes are now outputted, along with comparison across replicates

### Fixed
- Files are now outputted in a more manageable directory tree
- Code split into several files to improve maintainability

## 2016-01-16
### Added
- tRNA indexes now end with CCA modification
- iCLIPro is now run on output

### Fixed
- The 5' end of RT stops on the minus strand is now correctly reported (off-by-one error)

## 2015-12-26
### Added
- Both + and - signal are now reported for each nucleotide in PlotData_RepeatRNAHist_*
- Mean and standard deviation for each significant RT stop are now reported in files that end with _mergedRT.bed_stats

## 2015-12-25
### Added
- Added the ability to run fasticlip outside its installation directory once the appropriate environment variables are set

## 2015-12-23
### Added
- Can now handle any number of replicates. Added default thresholds for RT stop significance for different numbers of replicates.
- Users can define significance rules for RT stops for repeat and non-repeat RNAs separately with -tr and -tn

## 2015-12-16
### Added
- ViralGenome_analysis.py added; can now map to arbitrary viral genome Bowtie2 indexes

## 2015-09-24
### Added
- Can now specify filename prefix in motifAnalysis.py

### Fixed
- In Figure 5, if PlotData files are empty the plots will not be drawn (rather than crashing)

## 2015-09-23
### Fixed
- Clarified the definition of exons, CDS, introns, and UTRs in both fasticlip and motifAnalysis.py
- Added the file PlotData_ReadsPerGene_proteinCoding_byRegion to count up reads for each of the above regions for each gene
- Removed -f value from the bedtools intersect commands, since default is -f 1e-9 which should work

## 2015-09-20
### Added
- Users can now specify the minimum MAPQ quality they want to keep using the -q flag

## 2015-09-14
### Added
- mapping to tRNA indexes (docs.tar.gz also updated to add tRNA indexes)
- tRNA and repeat masker reads are now reported

### Fixed
- Default lower limit read length is 15
- Duplicate removal of reads is now done before 3'/5' trimming to save time
- Various redundancy fixes to speed up making Figures 1, 3, 4
- Using the -sorted flag in bedtools intersect to speed up the repeat and blacklist masking

## 2015-07-13
### Added
- Adding a list of genes that have 5' and 3' UTR binding, but no CDS binding

## 2015-06-12
### Added
- In motifAnalysis.py, lists of RT stops are now filtered so that there is only one representative per significant "window"

## 2015-05-27
### Fixed
- rDNA histogram is now the correct length

## 2015-04-12
### Fixed
- CLIP clusters are now merged if 1) they are duplicates of each other, or 2) they overlap

## 2015-02-26
### Changed
- changed location of docs.tar.gz and updated ./configure

## 2015-01-09
### Added
- added a script to map to retroviral indexes
- Added the necessary retroviral fasta files to the docs.tar.gz file

## 2014-12-30
### Added
- added motifAnalysis.py to analyze motifs using HOMER

## 2014-12-21
### Fixed
- we now combine all isoforms of each gene during plot making

## 2014-12-04
### Added
- Added a minimum read length post-5' trim

## 2014-11-09
### Fixed
- Changed -f13 to -f14 for the 5' trim so that the first 13 nucleotides are trimmed by default. -f should represent the first nucleotide that is kept













motifAnalysis.py
# motifAnalysis.py
# 12/30/14
# Motif_analysis.ipynb, in script form

import sys, os, re, cmath, math, glob, subprocess, csv, matplotlib, argparse
import numpy as np
import matplotlib.pyplot as plt 
from matplotlib.backends.backend_pdf import PdfPages
import pandas as pd

matplotlib.rcParams['savefig.dpi'] = 2 * matplotlib.rcParams['savefig.dpi']
csv.register_dialect("textdialect", delimiter='\t')

# *** INPUT : Sample to process. Please fill these in
# ---------------------------------------------------

### Parsing arguments ###

parser = argparse.ArgumentParser(description="motifAnalysis.py: use this to look for motifs in CLIP data", epilog="Example: python motifAnalysis.py -n ddx3wt --hg19 -l 12,14 -f 15 -m 25")
group = parser.add_mutually_exclusive_group()
parser.add_argument('-n', metavar='SAMPLENAME', help="Sample name; name of folder inside results/", required=True)
parser.add_argument('-p', metavar='PREFIX', help="Prefix for folders where motifs will be stored", default="homer")
group.add_argument('--hg19', action='store_true', help="required if your CLIP is from human")
group.add_argument('--mm9', action='store_true', help="required if your CLIP is from mouse")
parser.add_argument('-l', metavar='LENGTHS', help="Length(s) of motifs desired; separated by a ','. Default is '6,8,10,12'.", default="6,8,10,12")
parser.add_argument('-f', metavar='FILTER_WINDOW', help="Distance over which RT stops cannot overlap. Default is 10.", default=10)
parser.add_argument('-m', metavar='MOTIF_WINDOW', help="Size of window to find motifs around RT stop. Default is 20.", default=20)
parser.add_argument('--window', action='store_true', help="look at genes with reads in a certain window")
parser.add_argument('-s', metavar='START', type=int, help="Left end of window. 0 = left, 200 = 5UTR/CDS boundary, 400 = CDS/3UTR boundary, 600 = right. Default is 0.", default=0)
parser.add_argument('-e', metavar='END', type=int, help="Right end of window. Default is 600.", default=600)


# organism
args = parser.parse_args()
if not (args.hg19 or args.mm9):
	print "Error: must include --hg19 or --mm9. Exiting."
	exit()
if args.hg19: 
	org = 'hg19' 
	orgname='human'
else: 
	org = 'mm9'
	orgname = 'mouse'

sampleName=args.n #name of folder
if not glob.glob("results/" + sampleName):
	print "Folder %s does not exist. Exiting." %sampleName
	exit()
	
lengths=args.l #desired motif lengths, separated by commas with NO spaces
filter_win=int(args.f) #window around RT stop over which no other RT stops can be found
homer_win=int(args.m) #window around RT stop to find motifs within. if homer_win = 20 then we do pos +- 10
# Start/end of region to extract, between 0 and 300
if args.window:
	start=args.s
	end=args.e



# - Functions for extracting cluster data associated with gene list - 
# -------------------------------------------------------------------
def grep(pattern,filein):
	# Usage: Open a file and search all lines for a pattern.
	# Input: Pattern to search (gene name) and file name.
	# Output: List of lines in a file that have the pattern.
	r = []
	filein_open = open(filein, 'r')
	for line in filein_open:
		 # print line
		 if re.search(pattern,line):
			r.append(line)
	filein_open.close()
	return r

def extractClusters(geneList,allRT, filter_win):
	# Usage: Extract a set of CLIPper RT stops based on gene name
	# Input: Gene list
	# Output: RT stops in those genes
	# BD Update 12/13/2014: only unique RT stops
	# BD Update 6/11/2015: filterSummit.pl AND no more unique RT stops
	extractedRT = geneList+'_rtstops.bed'
	outfh = open(extractedRT, 'w')
	writer = csv.writer(outfh, 'textdialect')
	
	# List of allowed gene names (genes that have RT stops only in the defined region)
	namesToQuery = np.genfromtxt(geneList,usecols=(0,),delimiter='\t',dtype='string')
	
	# Iterate through all RT stops in clusters
	ifile = open(allRT, 'r')
	reader = csv.reader(ifile, 'textdialect')
	
	for row in reader:
		geneName = row[3].split('_')[0].split('.')[0]
		if geneName in namesToQuery: writer.writerow(row)
		
	ifile.close()
	outfh.close()
	
	# Run filtering
	proc = subprocess.Popen(["dos2unix",extractedRT])
	proc.communicate()
	filteredRT = geneList+'_rtstops_summits.bed'	
	proc = subprocess.Popen(["perl","filterSummit.pl",extractedRT, filteredRT, str(filter_win)])
	proc.communicate()	
	
	return filteredRT

# - Functions for running HOMER analysis - 
def shuffleBedFile(inBed):
	# Usage: Shuffle a bed file (usually used a background file for HOMER analysis) 
	# Input: Bedfile
	# Output: SHuffled bedFile
	program = 'shuffleBed'
	referenceFile = os.getcwd()+'/docs/%s/%s_transcriptome_collapse_exon.bed'%(org,org)
	genomeFile = os.getcwd()+'/docs/%s/%s.%s.genome'%(org, orgname, org)
	try:
		shuffledBed = inBed.replace('.bed','_shuffled.bed')
		outfh = open(shuffledBed, 'w')
		proc = subprocess.Popen([program,'-i',inBed,'-incl',referenceFile,'-g',genomeFile],stdout=outfh)
		proc.communicate()
		return shuffledBed
	except:
		print "Problem generating shuffled bedfile."
		
def makeBedForHOMER(inBed, homer_win):
	# Usage: This modified the bedfile for processing homer by making the first field a concatenation of chr_start_end
	# Input: Clean bedfile with first 5 field properly assigned
	# Output: Modified bed file with first field chr_start_end, and name excluded.
	# BD 6/11/2015: adds a +/-win to each RT stop identified as the summit of each cluster
	# Make sure bedfile only has 5 fields
	print inBed
	bedForHOMER=inBed.replace('.bed','_forHOMER.bed')	
	print bedForHOMER
	# Open new file
	f = open(bedForHOMER, 'w')
	with open(inBed, 'r') as infile:
		for line in infile:	
			elementList = line.strip().split('\t')
			if len(elementList) != 6: continue
			# Re-write the bed file with chr replaced
			f.write('\t'.join((elementList[0]+'_'+elementList[1]+'_'+elementList[2],elementList[0],str(int(elementList[1])-homer_win/2),str(int(elementList[2])+homer_win/2),elementList[5],'\n')))
	f.close()
	return bedForHOMER

def runHOMER(inBed,lengths,homer_win,outDirName):
	# Usage: Run the HOMER motif finding algorithm 
	# Input: Bedfile properly modified for HOMER
	# Output: A directory containing the HOMER output files
	program='findMotifsGenome.pl'
	homerReferenceFile = os.getcwd()+'/docs/%s/%s_transcriptome_collapse_exon.bed'%(org, org)
	# Convert the input bedFile into HOMER compatible format
	inBedForHOMER=makeBedForHOMER(inBed, homer_win)
	# Get the path of the input file 
	path,filename=os.path.split(inBedForHOMER)
	outDir=path+'/'+outDirName
	
	# print path, filename
	# return

	# Call HOMER, which will generate a directory of files
	proc = subprocess.Popen([program,inBedForHOMER,org,outDir,'-rna','-bg',homerReferenceFile, '-len', lengths, '-S', '10'])
	proc.communicate()
	

def convertENBLids(inNames):
	# Usage: Convert ENST to ENSG (unique ID) using ENSEMBL annotation file
	# Input: List of ENST IDs
	# Output: List of ENSG IDs
	genesFile = os.getcwd() + '/docs/%s/%s_ensembl_genes.txt'%(org, org)
	ensemblIDfile=np.genfromtxt(genesFile,usecols=(1,12,),delimiter='\t',dtype='string') # Note that column lookup is zero indexed
	temp=[]
	for name in inNames:
		outName=ensemblIDfile[ensemblIDfile[:,0]==name,1]
		temp=temp+[outName]
	temp=np.array(temp)
	return temp

def customGeneExtraction(outfilepath,startIndex,endIndex):
	# Extract genes with reads that fall within a specified window 
	# Input: File path, and coordinates for data selection
	# Output: None
	
	# Extract data from treat matrix
	treatMatrixCols=600
	treatMatrix=glob.glob(outfilepath+'/rawdata/*treatmatrix.txt')[0]
	treatMatrixData=np.genfromtxt(treatMatrix,skip_header=1,usecols=range(1,treatMatrixCols+1),delimiter='\t',dtype='float')
	geneNames=np.loadtxt(treatMatrix,dtype='string',skiprows=1,usecols=(0,),delimiter='\t')
	#print geneNames[0:10]
	# pathToNameConversion=os.getcwd() + '/docs/refSeq_to_Ensl_all.txt'
	# nameConversionToEnsembl=np.genfromtxt(pathToNameConversion,usecols=(0,1,),delimiter='\t',dtype='string') # Note that column lookup is zero indexed
	
	# Convert to ENSG IDs and check for genes in the initial list 
	geneNames=convertENBLids(geneNames)
	masterList = outfilepath+'/rawdata/clipGenes_proteinCoding'
	masterNames = np.genfromtxt(masterList,usecols=(0,),delimiter='\t',dtype='string') # Gene names isolated from Figure 1e
	indexer=[]
	for geneName in geneNames:
		if geneName in masterNames:
			indexer=indexer+[1]
		else:
			indexer=indexer+[0]
	indexer=np.array(indexer,dtype=bool)
	geneNames=geneNames[indexer]
	treatMatrixData=treatMatrixData[indexer,:]
	
	customData=treatMatrixData[treatMatrixData[:,range(startIndex,endIndex+1)].sum(axis=1) > 0,:]
	customNames=geneNames[treatMatrixData[:,range(startIndex,endIndex+1)].sum(axis=1) > 0]
	
	tosave=outfilepath+'TEMP'
	np.savetxt(tosave,customData,fmt="%s")
	
	treatSums=customData.sum(axis=1)
	sortedIndex=list(reversed([i[0] for i in sorted(enumerate(treatSums),key=lambda x:x[1])]))
	sortedData=customData[sortedIndex,:]
	sortedNames=customNames[sortedIndex]
	tosave=outfilepath+'SourceData_CustomGeneExtraction'
	np.savetxt(tosave,np.unique(sortedNames),fmt="%s")
	return tosave

def plotExtractedRegion(sample,plotNum,start,end):
	outfilepath=os.getcwd() + '/results/%s/'%sample
	# Bed file with protein coding reads
	filteredProteinCoding = outfilepath+'/rawdata/clipGenes_proteinCoding_LowFDRreads_centerCoord_snoRNAremoved_miRNAremoved.bed'
	averageGraph=outfilepath+'/rawdata/clipGenes_proteinCoding_LowFDRreads_centerCoord_snoRNAremoved_miRNAremoved_cleaned_sorted_UTRs_scaled_cds200_abt0_averageGraph.txt'	
	# Number of columns 
	avgGraphCols=600
	avgGraphData=np.loadtxt(averageGraph,skiprows=1,dtype='float',usecols=range(1,avgGraphCols+1))
	plt.subplot(1,2,plotNum)
	ylimit=max(avgGraphData[1,:])*1.1
	a= plt.plot(avgGraphData[1,:],color='blue',linewidth='1')
	plt.ylim(0,ylimit)
	plt.vlines(200,0,ylimit,linestyles='dashed')
	plt.vlines(400,0,175,linestyles='dashed')
	plt.tick_params(axis='x',labelbottom='off') 
	plt.axvspan(start,end,facecolor='r',alpha=0.5)
	plt.tick_params(axis='y',labelsize=5) 
	plt.title(sample)
	
	b=PdfPages("extractedRegion.pdf")
	b.savefig(a)
	b.close()
	
	
	
	
	
	
	
# Running things
# --------------
outfilepath=os.getcwd()+'/results/%s/'%sampleName
clipperClusters=glob.glob(outfilepath+'/rawdata/*threshold*mergedRT_CLIP_clusters_lowFDRreads_centerCoord_cleaned_sorted.bed')[0]
print "Cluster file to process: %s"%clipperClusters

if not args.window:
	# regions we want to look at
	UTRcluster_names = ['5p', '3p', 'cds', 'introns', 'exons']
	UTRclusters = ["clipGenes_proteinCoding_LowFDRreads_centerCoord_snoRNAremoved_miRNAremoved_" + x + '.bed' for x in UTRcluster_names]
	
	# - Run HOMER on all clusters  - 
	proc = subprocess.Popen(["dos2unix",clipperClusters])
	# proc.communicate()
	filteredRT = clipperClusters[:-4]+"rtstops_summits.bed"
	proc = subprocess.Popen(["perl","filterSummit.pl",clipperClusters, filteredRT, str(filter_win)])
	# proc.communicate()
	runHOMER(filteredRT,lengths,homer_win,'homer_allReads')
	# shuffledReads=shuffleBedFile(filteredRT)
	# runHOMER(shuffledReads,lengths,'homer_allReads_shuffle')

	# - Run HOMER on clusters for specific regions - 
	folderNames = [args.p + '_' + x for x in UTRcluster_names]
	i=0
	for clusterFile in UTRclusters:
		runHOMER(outfilepath + '/rawdata/' + clusterFile,lengths,homer_win,folderNames[i])
		i+=1
else:
	# Input: Extract all reads in the region of interest
	pathToCustomFile=customGeneExtraction(outfilepath,start,end)

	# Plot the region
	plotExtractedRegion(sampleName,1,start,end)

	# Run HOMER
	outfilepath=os.getcwd()+'/results/%s/'%sampleName
	clusters=extractClusters(pathToCustomFile,clipperClusters)
	runHOMER(clusters,lengths,'rawdata/homer_custom')